### ISSUE TITLE
Join field type throwing "current_type [join], merged_type [ObjectMapper]" in 7.1.0

### ISSUE THREAD

**Issue by ffknob** [elastic/elasticsearch#43012]
**Elasticsearch version**:
7.1.0

**Plugins installed**: []

**JVM version** (`java -version`):
openjdk version "10.0.2" 2018-07-17
OpenJDK Runtime Environment (build 10.0.2+13-suse-lp150.7.1-x8664)
OpenJDK 64-Bit Server VM (build 10.0.2+13-suse-lp150.7.1-x8664, mixed mode)

**OS version** (`uname -a` if on a Unix-like system):
Linux berlin 4.12.14-lp150.12.25-default #1 SMP Thu Nov 1 06:14:23 UTC 2018 (3fcf457) x86_64 x86_64 x86_64 GNU/Linux

**Description of the problem including expected versus actual behavior**:
I was trying to follow the examples provided by the documentation related to the `join` field type, but I could not reproduce them in 7.1.0. Although they work in 6.5.0.

**Steps to reproduce**:

 1. Create the index
```
PUT my-index
{
  "mappings": {
    "properties": {
      "text": {
        "type":  "text"
      },
      "join": {
        "type": "join",
        "relations": {
          "question": "answer"
        }
      } 
    }
  }
}
```
 2. Try to create a parent document:
```
POST my-index/1
{
  "text": "test",
  "join": {
    "name": "question"
  }
}
```
 Result:
```
{
  "error": {
    "root_cause": [
      {
        "type": "illegal_argument_exception",
        "reason": "mapper [join] of different type, current_type [join], merged_type [ObjectMapper]"
      }
    ],
    "type": "illegal_argument_exception",
    "reason": "mapper [join] of different type, current_type [join], merged_type [ObjectMapper]"
  },
  "status": 400
}
```

The same example (with `doc` as a type) work in 6.5.0.

**Provide logs (if relevant)**:

**Comment by ffknob** [comment]
I just got it to work by including the type `_doc` in the URL... Actually that's how the examples for the `join` datatype in 7.x are written, but I completely ignored it because I thought types were completely gone in 7.x...

```
POST my-index/_doc/1
{
  "text": "test",
  "join": "question"
}
``` 

Still a bug?

**Comment by jimczi** [comment]
In 7.x `POST my-index/1` means add a document in an index named `my_index` of type `1`. This should result in an error with a message like:
````
"Rejecting mapping update to [my-index] as the final mapping would have more than 1 type: [_doc, 1]"
````
but it seems that we check the field first even though we know that the mapping update will fail (because only one type is allowed per index). This should be considered as a bug because the error message is misleading but as you already noticed the correct syntax is to use `POST my-index/_doc/1`.

**Comment by jtibshirani** [comment]
I merged #45157 to address the confusing error message. Thanks @ffknob for filing this!
